src/main/java/org/quiltmc/loader/impl/config/Json5Serializer.java

```diff
@@ -0,0 +1,260 @@
+			writer.name(node.getKey().getLastComponent());
+			writer.beginObject();
+
+			for (ValueTreeNode child : ((ValueTreeNode.Parent) node)) {
+				serialize(writer, child);
+			}
+
+			writer.endObject();
+		} else {
+			TrackedValue<?> trackedValue = ((TrackedValue<?>) node);
+			Object defaultValue = trackedValue.getDefaultValue();
+
+			if (!(defaultValue instanceof CompoundConfigValue<?>)) {
+				writer.comment("default: " + defaultValue);
+			}
+
+			for (Constraint<?> constraint : trackedValue.constraints()) {
+				writer.comment(constraint.getRepresentation());
+			}
+
+			writer.name(node.getKey().getLastComponent());
+
+			serialize(writer, trackedValue.getRealValue());
+		}
+	}
+
+	@Override
+	public void serialize(Config config, OutputStream to) throws IOException {
+		JsonWriter writer = JsonWriter.json5(new OutputStreamWriter(to));
+
+		for (String comment : config.metadata(MetadataType.COMMENT)) {
+			writer.comment(comment);
+		}
+
+		writer.beginObject();
+
+		for (ValueTreeNode node : config.nodes()) {
+			this.serialize(writer, node);
+		}
+
+		writer.endObject();
+		writer.close();
+	}
+
+	@SuppressWarnings("unchecked")
+	private Object coerce(Object object, Object to) {
+		if (to instanceof Integer) {
+			return ((Number) object).intValue();
+		} else if (to instanceof Long) {
+			return ((Number) object).longValue();
+		} else if (to instanceof Float) {
+			return ((Number) object).floatValue();
+		} else if (to instanceof Double) {
+			return ((Number) object).doubleValue();
+		} else if (to instanceof String) {
+			return object;
+		} else if (to instanceof Boolean) {
+			return object;
+		} else if (to instanceof ValueMap) {
+			@SuppressWarnings("rawtypes")
+			ValueMap.Builder builder = ValueMap.builder(((ValueMap) to).getDefaultValue());
+
+			for (Map.Entry<String, ?> entry : ((Map<String, ?>) object).entrySet()) {
+				builder.put(entry.getKey(), this.coerce(entry.getValue(), ((ValueMap<?>) to).getDefaultValue()));
+			}
+
+			return builder.build();
+		} else if (to instanceof ValueList) {
+			Object[] values = ((List<?>) object).toArray();
+
+			for (int i = 0; i < values.length; ++i) {
+				values[i] = this.coerce(values[i], ((ValueList<?>) to).getDefaultValue());
+			}
+
+			return ValueList.create(((ValueList<?>) to).getDefaultValue(), values);
+		} else {
+			throw new RuntimeException();
+		}
+	}
+
+	@Override
+	@SuppressWarnings("unchecked")
+	public void deserialize(Config config, InputStream from) {
+		try {
+			JsonReader reader = JsonReader.json5(new InputStreamReader(from));
+
+			Map<String, Object> values = parseObject(reader);
+
+			for (TrackedValue<?> value : config.values()) {
+				Map<String, Object> m = values;
+
+				for (int i = 0; i < value.getKey().length(); ++i) {
+					String k = value.getKey().getKeyComponent(i);
+
+					if (m.containsKey(k) && i != value.getKey().length() - 1) {
+						m = (Map<String, Object>) m.get(k);
+					} else if (m.containsKey(k)) {
+						//noinspection rawtypes
```

Comment:
`noinspection` comments are exclusive to IDEA. This should be moved to a `@SuppressWarnings` annotation.